text/0032-upgrade-bootstrapper.md

```diff
@@ -0,0 +1,152 @@
+`casper-node`'s `config.toml`.
+
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+### Changes to existing `casper-node` (running in validator mode)
+
+We require the following changes to `casper-node`:
+
+* running in validator mode, when we reach the upgrade activation point, exit gracefully
+* running in validator mode, if the node starts, joins and finds that it has already reached the upgrade activation
+  point, exit gracefully
+* upgrade activation points should be defined in terms of era IDs, not block heights
+* config.toml updated to take the path to the chainspec upgrade file, similar to the `node.chainspec_config_path`
+  option.  Alternatively, just remove this option, and require the chainspec, its upgrade file, accounts.csv and the
+  config.toml file all to live in the same folder
+
+### Data migration
+
+We require to add a new `migrate-data` subcommand to `casper-node`.  We could provide a separate binary to be used
+exclusively for data and config migration, but it seems simpler to just make this a subcommand.  Simpler for the
+installer, for the `upgrade-bootstrapper` and simpler to code.
+
+Migration of any of the databases needs to take the following approach by iterating each value in the DB and doing:
+
+1. parse as the old type
+2. convert it to the new type
+3. store the new type, possibly under a different key
+4. if stored under a different key (i.e. it's not just an overwrite), then delete the old value
+
+We need to be able to resume migration in case it's killed while running.  This means we need to ensure that we either
+keep track of our progress by writing to disk after each individual item is successfully migrated, or else we ensure
+that parsing a new value as an old value fails, or else we ensure that running the conversion step on an
+already-converted item is a no-op.
+
+If data migration is going to take a prohibitively long time to execute, we should look to have the `upgrade-installer`
+perform _data_ migration only (not config migration) as soon as the installer provides the new `casper-node` binary,
+while the current `casper-node` is still running in validator mode.  However, this might result in a doubling or more of
+the disk space consumed by the databases, since the migrated data cannot be allowed to overwrite the existing data while
+the node is still running as a validator, and _this_ disk consumption might prove to be prohibitive.  Data migration
+would also still need to be executed again once the old `casper-node` stops due to reaching the upgrade activation
+point, since more data will likely have been written to the databases between then and when the data migration was
+previously completed.
+
+### Config migration
+
+Migration will also need to handle converting an old config file to a new one.  For this to work, we need to know which
```

Comment:
I am not sure this is necessary. Users are not allowed to modify chainspec, only the node configuration.

Reply:
It's the node configuration I'm meaning here, not the chainspec.

Reply:
cannot we copy over the old config and reuse it? if there are any new config parameters we can add them.

Reply:
Yes, but an issue is that we'd be hard-coding default values into the Rust code rather than providing defaults via config files as we currently do.  I think it's better to stick to how we do it currently to avoid changing too much.